### PR TITLE
MODDCB-90 Accept existing circulation request ID

### PR DESCRIPTION
**Pull request by MagzhanArtykov**
Covers: [MODDCB-90](https://folio-org.atlassian.net/browse/MODDCB-90)

## Purpose
For ECS TLR implementation, mod-dcb needs to be able to accept transactions containing an ID of an already existing circulation request.

### CODE REVIEW

File: src/main/java/org/folio/dcb/controller/EcsTlrTransactionApiController.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+@RestController
+@Log4j2
+@RequiredArgsConstructor
+public class EcsTlrTransactionApiController implements EcsRequestTransactionsApi {
+
+  private final EcsTlrTransactionsService ecsTlrTransactionsService;
+  private final TransactionAuditService transactionAuditService;
+
+  @Override
+  public ResponseEntity<TransactionStatusResponse> createEcsRequestTransactions(String ecsTlrTransactionId, DcbTransaction dcbTransaction) {
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/dcb/controller/EcsTlrTransactionApiController.java:23]: Please try to keep lines' length under 100 chars

File: src/main/java/org/folio/dcb/service/impl/EcsTlrTransactionsServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+@RequiredArgsConstructor
+@Log4j2
+public class EcsTlrTransactionsServiceImpl implements EcsTlrTransactionsService {
+
+  private final BaseLibraryService baseLibraryService;
+  private final TransactionRepository transactionRepository;
+
+  @Override
+  public TransactionStatusResponse createEcsTlrTransaction(String dcbTransactionId, DcbTransaction dcbTransaction) {
+    log.debug("createCirculationRequest:: creating new transaction request for role {} ", dcbTransaction.getRole());
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/dcb/service/impl/EcsTlrTransactionsServiceImpl.java:24]: I think we can log this in `info`
```suggestion
    log.info("createCirculationRequest:: creating new transaction request for role {} ", dcbTransaction.getRole());
```

File: src/main/java/org/folio/dcb/service/impl/EcsTlrTransactionsServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    checkEcsTransactionExistsAndThrow(dcbTransactionId);
+    if (dcbTransaction.getRole() == LENDER) {
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/dcb/service/impl/EcsTlrTransactionsServiceImpl.java:26]: These two branches are identical, aren't they?

File: src/main/java/org/folio/dcb/service/impl/EcsTlrTransactionsServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    }
+    return TransactionStatusResponse.builder()
+      .status(TransactionStatusResponse.StatusEnum.CREATED)
+      .item(dcbTransaction.getItem())
+      .patron(dcbTransaction.getPatron())
+      .build();
+  }
+
+  private void checkEcsTransactionExistsAndThrow(String dcbTransactionId) {
+    if(transactionRepository.existsById(dcbTransactionId)) {
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/dcb/service/impl/EcsTlrTransactionsServiceImpl.java:41]:
```suggestion
    if (transactionRepository.existsById(dcbTransactionId)) {
```

File: src/main/java/org/folio/dcb/service/impl/EcsTlrTransactionsServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+      throw new ResourceAlreadyExistException(
+        String.format("unable to create ecs transaction with id %s as it already exists", dcbTransactionId));
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/dcb/service/impl/EcsTlrTransactionsServiceImpl.java:43]:
```suggestion
        String.format("unable to create ECS transaction with ID %s as it already exists", dcbTransactionId));
```

File: descriptors/ModuleDescriptor-template.json [hunk 6]
```diff
@@ -122,6 +122,17 @@
...
             "circulation.check-in-by-barcode.post"
           ]
+        },
+        {
+          "methods": [
+            "POST"
+          ],
+          "pathPattern": "/ecs-tlr-transaction/{ecsTlrTransactionId}",
+          "permissionsRequired": [
+            "dcb.transactions.post"
```
[the remark this answers is not in the document]
Comment (MagzhanArtykov) [reply descriptors/ModuleDescriptor-template.json:132]: Changed to ecs-tlr.transactions.post

File: src/main/java/org/folio/dcb/service/impl/EcsTlrTransactionsServiceImpl.java (hunk 3 above)
Comment (MagzhanArtykov) [reply src/main/java/org/folio/dcb/service/impl/EcsTlrTransactionsServiceImpl.java:26]: I changed implementation. In future there will be different logic

File: src/main/java/org/folio/dcb/controller/EcsRequestTransactionsApiController.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+@Log4j2
+@RequiredArgsConstructor
+public class EcsRequestTransactionsApiController implements EcsRequestTransactionsApi {
+
+  private final EcsRequestTransactionsService ecsRequestTransactionsService;
+  private final TransactionAuditService transactionAuditService;
+
+  @Override
+  public ResponseEntity<TransactionStatusResponse> createEcsRequestTransactions(
+            String ecsRequestTransactionId, DcbTransaction dcbTransaction) {
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/dcb/controller/EcsRequestTransactionsApiController.java:24]: fix indentation

File: src/main/java/org/folio/dcb/service/EcsRequestTransactionsService.java [hunk 8]
```diff
@@ -0,0 +1,9 @@
+package org.folio.dcb.service;
+
+import org.folio.dcb.domain.dto.DcbTransaction;
+import org.folio.dcb.domain.dto.TransactionStatusResponse;
+
+public interface EcsRequestTransactionsService {
+  TransactionStatusResponse createEcsRequestTransactions(String ecsRequestTransactionsId,
+                                                         DcbTransaction dcbTransaction);
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/dcb/service/EcsRequestTransactionsService.java:8]: fix indentation

File: src/main/java/org/folio/dcb/service/impl/EcsRequestTransactionsServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+@Log4j2
+public class EcsRequestTransactionsServiceImpl implements EcsRequestTransactionsService {
+
+  private final BaseLibraryService baseLibraryService;
+  private final TransactionRepository transactionRepository;
+  private final CirculationRequestService circulationRequestService;
+
+  @Override
+  public TransactionStatusResponse createEcsRequestTransactions(String ecsRequestTransactionsId,
+                                                                DcbTransaction dcbTransaction) {
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/dcb/service/impl/EcsRequestTransactionsServiceImpl.java:30]: fix indentation

File: descriptors/ModuleDescriptor-template.json [hunk 10]
```diff
@@ -197,6 +216,19 @@
...
+      "displayName": "ECS request - all permissions",
+      "description": "All permissions for ECS request module",
+      "subPermissions": [
+        "ecs-request.transactions.post"
+      ]
+    },
+    {
+      "permissionName": "ecs-request.transactions.post",
+      "displayName": "creates new ecs request transaction",
+      "description": "creates new ecs request transaction"
```
Comment (OleksandrVidinieiev) [inline descriptors/ModuleDescriptor-template.json:231]:
```suggestion
      "displayName": "creates new ECS request transaction",
      "description": "creates new ECS request transaction"
```

File: src/main/java/org/folio/dcb/service/impl/EcsRequestTransactionsServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+        baseLibraryService.saveDcbTransaction(ecsRequestTransactionsId, dcbTransaction,
+          dcbTransaction.getRequestId());
+      } else {
+        throw new IllegalArgumentException("Unimplemented role: " + dcbTransaction.getRole());
+      }
+      return TransactionStatusResponse.builder()
+        .status(TransactionStatusResponse.StatusEnum.CREATED)
+        .item(dcbTransaction.getItem())
+        .patron(dcbTransaction.getPatron())
+        .build();
```
[the remark this answers is not in the document]
Comment (MagzhanArtykov) [reply src/main/java/org/folio/dcb/service/impl/EcsRequestTransactionsServiceImpl.java:59]: Most of code will be common for other role, so I moved role specific logic to separate method

**Review by OleksandrVidinieiev — APPROVED** [review]

**Review by roman-barannyk — APPROVED** [review]

**Pull request merged**